### PR TITLE
ARROW-16778: [C++] Fix build/unit test issues in msvc/win32

### PR DESCRIPTION
**Pull request by avertleyb**

### CODE REVIEW

File: cpp/src/arrow/util/bit_util.h [hunk 1]
```diff
@@ -67,7 +67,14 @@ static constexpr uint8_t kBytePopcount[] = {
     5, 4, 5, 5, 6, 3, 4, 4, 5, 4, 5, 5, 6, 4, 5, 5, 6, 5, 6, 6, 7, 3, 4, 4, 5, 4, 5, 5, 6,
     4, 5, 5, 6, 5, 6, 6, 7, 4, 5, 5, 6, 5, 6, 6, 7, 5, 6, 6, 7, 6, 7, 7, 8};
 
-static inline uint64_t PopCount(uint64_t bitmap) { return ARROW_POPCOUNT64(bitmap); }
+static inline uint64_t PopCount(uint64_t bitmap) {
+#if defined(_MSC_VER) && !defined(_M_AMD64) && !defined(_M_X64) 
+  const uint32_t* p = reinterpret_cast<const uint32_t*>((void*)(&bitmap));
+  return ARROW_POPCOUNT32(*p) + ARROW_POPCOUNT32(*(p + 1));
```
Comment (kou) [inline cpp/src/arrow/util/bit_util.h:73]: I'm not sure which is better for performance but it seems that we can use bit operations to split `uint64_t` value to `uint32_t` values:

```c++
  return ARROW_POPCOUNT32((bitmap >> 32) & UINT32_MAX) +
         ARROW_POPCOUNT32(bitmap & UINT32_MAX);
```

**Review by kou — COMMENTED** [review]
Could you also add a CI job for building 32bit binaries with Visual C++?

The following patch will work:

```diff
diff --git a/.github/workflows/cpp.yml b/.github/workflows/cpp.yml
index 82e99160c3..6eb2213c93 100644
--- a/.github/workflows/cpp.yml
+++ b/.github/workflows/cpp.yml
@@ -192,12 +192,15 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os:
-          - windows-2019
         include:
           - os: windows-2019
-            name: Windows 2019
+            name: Windows 2019 x64
             generator: Visual Studio 16 2019
+            platform: x64
+          - os: windows-2019
+            name: Windows 2019 Win32
+            generator: Visual Studio 16 2019
+            platform: Win32
     env:
       ARROW_BOOST_USE_SHARED: OFF
       ARROW_BUILD_BENCHMARKS: ON
@@ -222,7 +225,7 @@ jobs:
       ARROW_WITH_ZLIB: ON
       ARROW_WITH_ZSTD: ON
       BOOST_SOURCE: BUNDLED
-      CMAKE_ARGS: '-A x64 -DOPENSSL_ROOT_DIR=C:\Program Files\OpenSSL-Win64'
+      CMAKE_ARGS: '-A ${{ matrix.platform }} -DOPENSSL_ROOT_DIR=C:\Program Files\OpenSSL-Win64'
       CMAKE_GENERATOR: ${{ matrix.generator }}
       CMAKE_INSTALL_LIBDIR: bin
       CMAKE_INSTALL_PREFIX: /usr
```

File: cpp/src/arrow/util/bit_util.h (hunk 1 above)
Comment (avertleyb) [reply cpp/src/arrow/util/bit_util.h:73]: If performance is critical, I would assume this version would just add a couple of bit operation.  But if you feel it looks cleaner I can change this.  Please let me know.

Comment (avertleyb) [reply cpp/src/arrow/util/bit_util.h]: @kou updated PR

**Comment by avertleyb** [comment]
I just found out that I need to follow certain process while submitting fixes to open source projects.

I will re-submit these changes soon.

**Comment by jmckenna** [comment]
I see that the win32 workflow added here, is now removed from master (this would explain why users cannot compile arrow per https://github.com/apache/arrow/issues/32111 ).  Can we add this workflow back into master?

**Comment by kou** [comment]
This wasn't merged.
Do you want to work on this (#32111)?

**Pull request closed without merging**